### PR TITLE
gtsam: init at 4.2

### PR DESCRIPTION
**Pull request by fagg**
## Description of changes

Add gtsam @ version 4.2. This is my first nix package, any help appreciated. Thanks.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by AndersonTorres** [comment]
1. read contrib guide
2. do you understand Git?
3. the commit adding you as maintainer should be worded `maintainers: add fagg`
4. the commit adding a new package should be worded as `<package name>: init at <version>`

**Comment by fagg** [comment]
> 1. read contrib guide
> 2. do you understand Git?
> 3. the commit adding you as maintainer should be worded `maintainers: add fagg`
> 4. the commit adding a new package should be worded as `<package name>: init at <version>`

1. I did
2. Yes. Professional software engineer. 
3. & 4. Don’t these get squashed prior to merge?

File: pkgs/by-name/gt/gtsam/package.nix [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+, metis
+}:
+
+stdenv.mkDerivation rec {
+  pname = "gtsam";
+  version = "4.2";
+
+  src = fetchurl {
+    url = "https://github.com/borglab/gtsam/archive/refs/tags/4.2.tar.gz";
+    sha256 = "9ff8846d0a83a245c284cb5760ec2d74535ef9b5885183ccfefd7ff122eba60e";
```
Comment (WolfangAukang) [inline pkgs/by-name/gt/gtsam/package.nix:17]:
```suggestion
    hash = "9ff8846d0a83a245c284cb5760ec2d74535ef9b5885183ccfefd7ff122eba60e";
```

File: pkgs/by-name/gt/gtsam/package.nix [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    "-DGTSAM_USE_SYSTEM_METIS=ON"
+    "-DGTSAM_SLOW_BUT_CORRECT_BETWEENFACTOR=OFF"
+  ];
+
+  doCheck = true;
+
+  meta = with lib; {
+    description = "Library for sensor fusion with factor graphs and Bayes networks";
+    homepage = "https://gtsam.org";
+    licences = licenses.bsd3;
```
Comment (WolfangAukang) [inline pkgs/by-name/gt/gtsam/package.nix:67]:
```suggestion
    licenses = licenses.bsd3;
```

File: pkgs/by-name/gt/gtsam/package.nix [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+, eigen
+, metis
+}:
+
+stdenv.mkDerivation rec {
+  pname = "gtsam";
+  version = "4.2";
+
+  src = fetchurl {
+    url = "https://github.com/borglab/gtsam/archive/refs/tags/4.2.tar.gz";
```
Comment (WolfangAukang) [inline pkgs/by-name/gt/gtsam/package.nix:16]:
```suggestion
    url = "https://github.com/borglab/gtsam/archive/refs/tags/${version}.tar.gz";
```

File: pkgs/by-name/gt/gtsam/package.nix [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  ];
+
+  doCheck = true;
+
+  meta = with lib; {
+    description = "Library for sensor fusion with factor graphs and Bayes networks";
+    homepage = "https://gtsam.org";
+    licences = licenses.bsd3;
+    maintainers = with maintainers; [ fagg ];
+    platforms = platforms.linux;
```
Comment (WolfangAukang) [inline pkgs/by-name/gt/gtsam/package.nix:69]:
```suggestion
    platforms = platforms.linux;
    changelog = "https://github.com/borglab/gtsam/releases/tag/${src.rev}";
```

**Review by WolfangAukang — CHANGES_REQUESTED** [review]
Were you able to build this from source? I see the process according to the documentation is straightforward, so it could be worthy to try it.

**Comment by fagg** [comment]
> Were you able to build this from source? I see the process according to the documentation is straightforward, so it could be worthy to try it.

Yes.

File: pkgs/by-name/gt/gtsam/package.nix [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+, boost
+, eigen
+, metis
+}:
+
+stdenv.mkDerivation rec {
+  pname = "gtsam";
+  version = "4.2";
+
+  src = fetchurl {
```
Comment (kirillrdy) [inline pkgs/by-name/gt/gtsam/package.nix:15]: please use fetchFromGithub

**Comment by kirillrdy** [comment]
@fagg commits don't get squashed when merged. nixpkgs relies on commit message formatting for CI, so all PR's are required to follow format outlined in [CONTRIBITION.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md)

**Comment by fagg** [comment]
> @fagg commits don't get squashed when merged. nixpkgs relies on commit message formatting for CI, so all PR's are required to follow format outlined in [CONTRIBITION.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md)

Thank you.

All comments/suggestions addressed, and commit messages are fixed.

File: pkgs/by-name/gt/gtsam/package.nix [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+}:
+
+stdenv.mkDerivation rec {
+  pname = "gtsam";
+  version = "4.2";
+
+  src = fetchFromGitHub {
+    owner = "borglab";
+    repo = "gtsam";
+    rev = "${version}";
```
Comment (AndersonTorres) [inline pkgs/by-name/gt/gtsam/package.nix:18]:
```suggestion
    rev = version;
```

File: pkgs/by-name/gt/gtsam/package.nix [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    eigen
+    metis
+  ];
+
+  nativeBuildInputs = [
+    cmake
+  ];
+
+  cmakeFlags = [
+    "-DCMAKE_BUILD_TYPE=Release"
```
Comment (AndersonTorres) [inline pkgs/by-name/gt/gtsam/package.nix:41]: Use the lib.cmake* functions (you can find them on lib/strings.nix).

File: pkgs/by-name/gt/gtsam/package.nix [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    "-DGTSAM_SUPPORT_NESTED_DISSECTION=ON"
+    "-DGTSAM_USE_TANGENT_PREINTEGRATION=ON"
+    "-DGTSAM_USE_SYSTEM_METIS=ON"
+    "-DGTSAM_SLOW_BUT_CORRECT_BETWEENFACTOR=OFF"
+  ];
+
+  doCheck = true;
+
+  meta = with lib; {
+    description = "Library for sensor fusion with factor graphs and Bayes networks";
```
Comment (AndersonTorres) [inline pkgs/by-name/gt/gtsam/package.nix:67]: nit: asciibetical order

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/by-name/gt/gtsam/package.nix [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    homepage = "https://gtsam.org";
+    licenses = licenses.bsd3;
```
Comment (kirillrdy) [inline pkgs/by-name/gt/gtsam/package.nix:69]: - key 'meta.licenses' is unrecognized; expected one of:
         ['available', 'badPlatforms', 'branch', 'broken', 'changelog', 'description', 'downloadPage', 'executables', 'homepage', 'hydraPlatforms', 'insecure', 'isBuildPythonPackage', 'isFcitxEngine', 'isGutenprint', 'isHydraChannel', 'isIbusEngine', 'knownVulnerabilities', 'license', 'longDescription', 'mainProgram', 'maintainers', 'maxSilent', 'name', 'outputsToInstall', 'pkgConfigModules', 'platforms', 'position', 'priority', 'schedulingPriority', 'sourceProvenance', 'tag', 'tests', 'timeout', 'unfree', 'unsupported', 'version']
       '

Comment (AndersonTorres) [inline pkgs/by-name/gt/gtsam/package.nix:69]:
```suggestion
    license = licenses.bsd3;
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
PING

File: pkgs/by-name/gt/gtsam/package.nix [hunk 10]
```diff
@@ -0,0 +1,74 @@
+{ lib
+, stdenv
```
Comment (FliegendeWurst) [inline pkgs/by-name/gt/gtsam/package.nix:2]: Please format with the new formatter.

File: pkgs/by-name/gt/gtsam/package.nix (hunk 6 above)
Comment (FliegendeWurst) [inline pkgs/by-name/gt/gtsam/package.nix:18]:
```suggestion
    rev = "refs/tags/${version}";
```

**Review by FliegendeWurst — CHANGES_REQUESTED** [review]

**Pull request closed without merging**